### PR TITLE
Reconcile restart failed envoy pods

### PR DESCRIPTION
**Pull request by emosbaugh**
In some scenarios with resource over utilization we are seeing envoy pods fall into an unrecoverable state after [shutting down and draining envoy](https://github.com/projectcontour/contour/blob/e067e1c59b44cd7b859b6ca8b958bd0018edc789/cmd/contour/shutdownmanager.go#L145-L147). See https://github.com/projectcontour/contour/issues/3192

This attempts to workaround the issue by forcefully deleting envoy pods that have fallen into this state for at least EnvoyPodsNotReadyDuration.

Example describe envoy pod in this state:

```
Name:         envoy-5z48v
Namespace:    projectcontour
...
Status:       Running
IP:           10.32.0.1
IPs:
  IP:           10.32.0.1
Controlled By:  DaemonSet/envoy
Init Containers:
  envoy-initconfig:
...
Containers:
  shutdown-manager:
    Container ID:  docker://3bb4827d7752253218b811b5c71c57e008065d84964bb1bf3ff492761a980112
    Image:         projectcontour/contour:v1.7.0
...
    State:          Running
      Started:      Fri, 29 Jul 2022 18:01:19 -0700
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Mon, 18 Jul 2022 06:06:40 -0700
      Finished:     Fri, 29 Jul 2022 18:01:19 -0700
    Ready:          True
    Restart Count:  18
    Liveness:       http-get http://:8090/healthz delay=3s timeout=1s period=10s #success=1 #failure=3
...
  envoy:
    Container ID:  docker://4af9db039c28209bd0174be92028a9517b528b28be01782cfb0fd65aefbd46c8
    Image:         envoyproxy/envoy:v1.15.0
...
    State:          Running
      Started:      Mon, 18 Jul 2022 06:06:40 -0700
    Last State:     Terminated
      Reason:       Error
      Exit Code:    255
      Started:      Mon, 18 Jul 2022 06:05:34 -0700
      Finished:     Mon, 18 Jul 2022 06:06:21 -0700
    Ready:          False
    Restart Count:  10
    Readiness:      http-get http://:8002/ready delay=3s timeout=1s period=4s #success=1 #failure=3
...
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
...
Events:
  Type     Reason     Age                     From     Message
  ----     ------     ----                    ----     -------
  Warning  Unhealthy  4m9s (x29475 over 32h)  kubelet  Readiness probe failed: HTTP probe failed with statuscode: 503
```

### CODE REVIEW

**Comment by emosbaugh** [comment]
[sc-37139]

File: pkg/cluster/contour.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		case "envoy":
+			if !status.Ready &&
+				status.State.Running != nil &&
+				status.State.Running.StartedAt.Time.Before(before) {
+				envoyNotReady = true
+			}
+		}
+	}
+
+	if !(shutdownManagerReady && envoyNotReady) {
```
Comment (ricardomaraschini) [inline pkg/cluster/contour.go:90]: This `if` took me a while to understand. Not sure if it comes from the fact that one variables has a "Not" on its name.

File: pkg/cluster/contour_test.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+					Type:               corev1.ContainersReady,
+					Status:             corev1.ConditionTrue,
+					LastTransitionTime: metav1.Time{Time: time.Now().Add(-time.Hour)},
+				}},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := shouldRestartEnvoyPod(tt.podStatus); got != tt.want {
```
Comment (ricardomaraschini) [inline pkg/cluster/contour_test.go:117]: I changed the test here to also include the missing argument for `shouldRestartEnvoyPod()` but yet the test fails.

File: pkg/cluster/clear_node.go [hunk 3]
```diff
@@ -31,10 +31,7 @@ func (c *Controller) ClearNode(ctx context.Context, nodeName string) error {
 			continue
 		}
 		c.Log.Infof("Force deleting pod %s/%s on node %s", pod.Namespace, pod.Name, nodeName)
-		var grace int64 = 0
-		err := c.Config.Client.CoreV1().Pods(pod.Namespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{
-			GracePeriodSeconds: &grace,
-		})
+		err := c.Config.Client.CoreV1().Pods(pod.Namespace).Delete(context.TODO(), pod.Name, *metav1.NewDeleteOptions(0))
```
[the remark this answers is not in the document]
Comment (emosbaugh) [reply pkg/cluster/clear_node.go:34]: will make this change

**Review by ricardomaraschini — APPROVED** [review]

**Pull request merged**